drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2023 EPAM Systems
+ * Copyright (c) 2023 IoT.bzh
+ *
+ * r8a779f0 Clock Pulse Generator / Module Standby and Software Reset
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT renesas_r8a779f0_cpg_mssr
+
+#include <errno.h>
+#include <zephyr/arch/cpu.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/renesas_cpg_mssr.h>
+#include <zephyr/dt-bindings/clock/renesas_cpg_mssr.h>
+#include <zephyr/dt-bindings/clock/r8a779f0_cpg_mssr.h>
+#include <zephyr/irq.h>
+#include "clock_control_renesas_cpg_mssr.h"
+
+#define LOG_LEVEL CONFIG_CLOCK_CONTROL_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(clock_control_rcar);
+
+struct r8a779f0_cpg_mssr_cfg {
+	DEVICE_MMIO_ROM; /* Must be first */
+};
+
+struct r8a779f0_cpg_mssr_data {
+	struct rcar_cpg_mssr_data cmn; /* Must be first */
+};
+
+/* NOTE: the array MUST be sorted by module field */
+static struct cpg_clk_info_table core_props[] = {
+	RCAR_CORE_CLK_INFO_ITEM(R8A779F0_CLK_S0D12_PER, RCAR_CPG_NONE,
+				RCAR_CPG_NONE, RCAR_CPG_KHZ(66660)),
+
+	RCAR_CORE_CLK_INFO_ITEM(R8A779F0_CLK_CL16M, RCAR_CPG_NONE,
+				RCAR_CPG_NONE, RCAR_CPG_KHZ(16660)),
+};
+
+/* NOTE: the array MUST be sorted by module field */
+static struct cpg_clk_info_table mod_props[] = {
+	RCAR_MOD_CLK_INFO_ITEM(702, R8A779F0_CLK_S0D12_PER),
+	RCAR_MOD_CLK_INFO_ITEM(704, R8A779F0_CLK_S0D12_PER),
+
+	RCAR_MOD_CLK_INFO_ITEM(915, R8A779F0_CLK_CL16M),
+};
+
+static int r8a779f0_cpg_enable_disable_core(const struct device *dev,
+					   struct cpg_clk_info_table *clk_info,
+					   uint32_t enable)
+{
+	ARG_UNUSED(enable);
+
+	int ret = 0;
+	uint32_t reg;
+
+	/* Not any used clocks needs to be manually endisable for now */
+	switch (clk_info->module) {
+	default:
+		ret = -ENOTSUP;
+		break;
+	}
+
+	if (!ret) {
+		rcar_cpg_write(DEVICE_MMIO_GET(dev), clk_info->offset, reg);
+	}
+	return ret;
+}
+
+static int r8a779f0_cpg_core_clock_endisable(const struct device *dev,
+					    struct rcar_cpg_clk *clk,
+					    bool enable)
+{
+	struct cpg_clk_info_table *clk_info;
+	struct r8a779f0_cpg_mssr_data *data = dev->data;
+	k_spinlock_key_t key;
+
+	clk_info = rcar_cpg_find_clk_info_by_module_id(dev, clk->domain, clk->module);
+	if (!clk_info) {
+		return -EINVAL;
+	}
+
+	if (enable) {
+		if (clk->rate > 0) {
+			int ret;
+			uintptr_t rate = clk->rate;
+
+			ret = rcar_cpg_set_rate(dev, (clock_control_subsys_t)clk,
+						(clock_control_subsys_rate_t)rate);
+			if (ret < 0) {
+				return ret;
+			}
+		}
+	}
+
+	key = k_spin_lock(&data->cmn.lock);
+	r8a779f0_cpg_enable_disable_core(dev, clk_info, enable);
+	k_spin_unlock(&data->cmn.lock, key);
+
+	return 0;
+}
+
+int r8a779f0_cpg_mssr_start_stop(const struct device *dev,
+	clock_control_subsys_t sys, bool enable)
```

Comment:
sorry one more indentation nitpick: under the previous parenthesis is how it's normally done here, you are mixing thigs up a bit in the file, here you are just doing a tab, in `r8a779f0_cpg_mssr_start` you are back one charcter, could you check these and make sure they are at least coherent within the file?

```
int r8a779f0_cpg_mssr_start_stop(const struct device *dev,
                                 clock_control_subsys_t sys, bool enable)
static int r8a779f0_cpg_mssr_start(const struct device *dev,
                                   clock_control_subsys_t sys)
```

is how it should be

Reply:
running clang-format on it as it is a new file could be the right call ?

Reply:
clang-format mostly works but often does a bit of a disaster on complex macros, if you want to use it fine but don't take the output religiously, either commit locally, clang-format and then manually `checkout -p` the messy chunks, or add some `/* clang-format off */` `/* clang-format on */` markers around the part it messes up, if any.